### PR TITLE
Add complex support for torch.mean [CUDA]

### LINKED ISSUE #46982: Mean_cuda not implemented for complex types
## 🐛 Bug

Tensor averaging not implemented for complex types

## To Reproduce

``` python
>>> import torch
>>> x = torch.ones(4, dtype=torch.complex128).cuda()
>>> torch.mean(x)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-108-34af2dc3edea> in <module>
      1 import torch
      2 x = torch.ones(4, dtype=torch.complex128).cuda()
----> 3 torch.mean(x)

RuntimeError: "mean_cuda" not implemented for 'ComplexDouble'
```

Same for float, `torch.complex64`, `torch.complex32`.

## Expected behaviour

Calculates mean across tensor, e.g.
``` python
tensor(1., device='cuda:0', dtype=torch.complex128)
```

## Environment

PyTorch version: 1.8.0.dev20201027
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: 10.0.1 
CMake version: version 3.18.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 970
Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.2
/usr/lib/libcudnn_adv_infer.so.8.0.2
/usr/lib/libcudnn_adv_train.so.8.0.2
/usr/lib/libcudnn_cnn_infer.so.8.0.2
/usr/lib/libcudnn_cnn_train.so.8.0.2
/usr/lib/libcudnn_ops_infer.so.8.0.2
/usr/lib/libcudnn_ops_train.so.8.0.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0.dev20201027
[pip3] torchvision==0.9.0.dev20201027
[pip3] torchviz==0.0.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] torch                     1.8.0.dev20201026          pypi_0    pypi
[conda] torchvision               0.9.0.dev20201027          pypi_0    pypi
[conda] torchviz                  0.0.1                    pypi_0    pypi

cc @ezyang @anjali411 @dylanbespalko @mruberry

### PR DESCRIPTION
**Pull request by RockingJavaBean**
Fixes #46982

### CODE REVIEW

File: aten/src/ATen/native/cuda/ReduceMomentKernel.cu [hunk 1]
```diff
@@ -36,8 +36,8 @@ static void std_var_kernel_cuda(TensorIterator& iter, bool unbiased, bool take_s
 
 template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
 void mean_kernel_impl(TensorIterator& iter) {
-  float factor = float(iter.num_output_elements()) / iter.numel();
-  gpu_reduce_kernel<scalar_t, out_t>(iter, MeanOps<acc_t, float> {factor});
+  scalar_t factor = scalar_t(iter.num_output_elements()) / iter.numel();
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/ReduceMomentKernel.cu:39]: what is this change for? I think factor should be float for all cases regardless of the scalar type

Comment (RockingJavaBean) [reply aten/src/ATen/native/cuda/ReduceMomentKernel.cu:45]: @anjali411 This change is to handle the complex tensors.
If we make the `factor` float for all dtypes, there would be compiling errors in the `project` function as no overloaded versions of  `*` operator are found between `complex<double>` and `float` operands.
https://github.com/pytorch/pytorch/blob/eec201c1388434a8d1efbfaf3d6ab091247b9aa9/aten/src/ATen/native/SharedReduceOps.h#L142-L144

Compiling logs
```
E:/repo/ml/github/pytorch/aten/src\ATen/native/SharedReduceOps.h(143): error: no operator "*" matches these operands
            operand types are: c10::complex<double> * const float
          detected during:
            instantiation of "acc_t at::native::MeanOps<acc_t, factor_t>::project(acc_t) const [with acc_t=c10::complex<double>, factor_t=float]"
```

**Comment by RockingJavaBean** [comment]
@anjali411 Thanks so much for reviewing this PR.
The latest PR leverages the `explicit template specialization` to handle `factor` with complex tensors, instead of casting `factor` to `scalar_t`.
The CI checks all passed, please kindly help review.

File: aten/src/ATen/native/cuda/ReduceMomentKernel.cu [hunk 2]
```diff
@@ -36,26 +36,40 @@ static void std_var_kernel_cuda(TensorIterator& iter, bool unbiased, bool take_s
 
 template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
 void mean_kernel_impl(TensorIterator& iter) {
-  float factor = float(iter.num_output_elements()) / iter.numel();
+  float factor = static_cast<float>(iter.num_output_elements()) / iter.numel();
   gpu_reduce_kernel<scalar_t, out_t>(iter, MeanOps<acc_t, float> {factor});
 }
 
+template <>
+void mean_kernel_impl<c10::complex<float>>(TensorIterator& iter) {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/ReduceMomentKernel.cu:44]: I think we can replace these two functions with:
```
template <typename T>
void mean_kernel_impl<c10::complex<T>>(TensorIterator& iter) {
  T factor = static_cast<T>(iter.num_output_elements()) / iter.numel();
  gpu_reduce_kernel<c10::complex<T>, c10::complex<T>>
    (iter, MeanOps<c10::complex<T>, T> {factor});
}
```

File: aten/src/ATen/native/cuda/ReduceMomentKernel.cu [hunk 3]
```diff
@@ -36,26 +36,40 @@ static void std_var_kernel_cuda(TensorIterator& iter, bool unbiased, bool take_s
 
 template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/ReduceMomentKernel.cu:37]: setting `acc_t = typename c10::scalar_value_type<scalar_t>::type` should resolve the issue here

`c10::scalar_value_type<scalar_t>::type` returns `scalar_t` for all non-complex dtypes and returns `T` for `c10::complex<T>`.

**Comment by anjali411** [comment]
> @anjali411 Thanks so much for reviewing this PR.
> The latest PR leverages the `explicit template specialization` to handle `factor` with complex tensors, instead of casting `factor` to `scalar_t`.

Hi @RockingJavaBean I think we shouldn't need to define overload functions for complex types, after the change I suggested in my comment. But this PR looks great overall, and should be ready to merge after that change!

**Review by anjali411 — APPROVED** [review]

File: test/test_torch.py [hunk 4]
```diff
@@ -514,7 +514,8 @@ def test_mean_dim(self):
             self._test_dim_ops(
                 lambda t, d: t.mean(d),
                 lambda n, d: n.mean(d),
-                use_integral=False)
+                use_integral=False,
+                use_complex=True)
```
Comment (anjali411) [inline test/test_torch.py:518]: This test doesn't run on CUDA. can you please extend the test for mean in `tensor_op_tests` to also test complex dtypes?

**Review by anjali411 — COMMENTED** [review]
let's update the CUDA test for mean to test complex dtypes as well

File: aten/src/ATen/native/cuda/ReduceMomentKernel.cu (hunk 3 above)
Comment (RockingJavaBean) [reply aten/src/ATen/native/cuda/ReduceMomentKernel.cu:37]: Thanks so much for the tip. The latest code now manipulates `c10::scalar_value_type<acc_t>::type` to get the type of the factor, the overload functions for complex numbers are not needed.

File: test/test_torch.py (hunk 4 above)
Comment (RockingJavaBean) [reply test/test_torch.py:518]: Thanks for pointing this out, the tests for complex dtypes are added to `tensor_op_tests`.

**Comment by RockingJavaBean** [comment]
@anjali411 I'm really grateful for your tip on `c10::scalar_value_type`, and the tests for complex inputs of `torch.mean` function are updated.
Please kindly help review.

**Review by anjali411 — APPROVED** [review]
lgtm and the windows test failure is an upstream test failure

**Comment by RockingJavaBean** [comment]
@anjali411 thanks so much for reviewing this PR, the CUDA tests for `torch.mean` passed before the upstream failure CI check terminated according to the `circleci` log.
```
test_mean_cuda_bfloat16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.005s)
test_mean_cuda_complex128 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_cuda_complex64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_cuda_float16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_cuda_float32 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.003s)
test_mean_cuda_float64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_dim_cuda_bfloat16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_dim_cuda_complex128 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.005s)
test_mean_dim_cuda_complex64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_dim_cuda_float16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_dim_cuda_float32 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_dim_cuda_float64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.003s)
test_mean_neg_dim_cuda_bfloat16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.005s)
test_mean_neg_dim_cuda_complex128 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_neg_dim_cuda_complex64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_neg_dim_cuda_float16 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_neg_dim_cuda_float32 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.004s)
test_mean_neg_dim_cuda_float64 (__main__.TestTensorDeviceOpsCUDA) ... ok (0.003s
```

**Comment by anjali411** [comment]
@RockingJavaBean can you please rebase?

**Comment by RockingJavaBean** [comment]
@anjali411 thank you so much for the kind reminder, I just rebased this PR with the latest code.

**Pull request closed without merging**